javaslang-benchmark/src/test/java/javaslang/benchmark/collection/TreeMapBenchmark.java

```diff
@@ -0,0 +1,398 @@
+ * /___/\_/  \_/\____/\_/  \_/\__\/__/\__\_/  \_//  \__/\_____/   Licensed under the Apache License, Version 2.0
+ */
+package javaslang.benchmark.collection;
+
+import javaslang.Tuple;
+import javaslang.Tuple2;
+import javaslang.benchmark.BenchmarkPerformanceReporter;
+import javaslang.collection.Array;
+import javaslang.collection.List;
+import javaslang.collection.SortedMap;
+import javaslang.collection.Stack;
+import javaslang.collection.TreeMap;
+import org.junit.Test;
+import org.openjdk.jmh.annotations.Benchmark;
+import org.openjdk.jmh.annotations.BenchmarkMode;
+import org.openjdk.jmh.annotations.Fork;
+import org.openjdk.jmh.annotations.Group;
+import org.openjdk.jmh.annotations.Level;
+import org.openjdk.jmh.annotations.Measurement;
+import org.openjdk.jmh.annotations.Mode;
+import org.openjdk.jmh.annotations.OutputTimeUnit;
+import org.openjdk.jmh.annotations.Param;
+import org.openjdk.jmh.annotations.Scope;
+import org.openjdk.jmh.annotations.Setup;
+import org.openjdk.jmh.annotations.State;
+import org.openjdk.jmh.annotations.TearDown;
+import org.openjdk.jmh.annotations.Threads;
+import org.openjdk.jmh.annotations.Warmup;
+import org.openjdk.jmh.results.RunResult;
+import org.openjdk.jmh.runner.Runner;
+import org.openjdk.jmh.runner.options.Options;
+import org.openjdk.jmh.runner.options.OptionsBuilder;
+import scala.collection.Iterator;
+import scala.math.Ordering;
+import scala.math.Ordering$;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Map;
+import java.util.NavigableMap;
+import java.util.Random;
+import java.util.concurrent.TimeUnit;
+
+@BenchmarkMode(Mode.Throughput)
+@Warmup(iterations = 2, time = 1, timeUnit = TimeUnit.SECONDS)
+@Measurement(iterations = 5, time = 1, timeUnit = TimeUnit.SECONDS)
+@OutputTimeUnit(TimeUnit.SECONDS)
+@Fork(value = 1, jvmArgsAppend = { "-XX:+UseG1GC", "-Xss500m", "-Xms4g", "-Xmx4g", "-ea" })
+@Threads(1)
+public class TreeMapBenchmark {
+    private static final Ordering<Integer> SCALA_ORDERING = Ordering$.MODULE$.comparatorToOrdering(Integer::compareTo);
+
+    @Test
+    public void launchBenchmark() throws Exception {
+        Options opt = new OptionsBuilder()
+                .include(this.getClass().getSimpleName())
+                .shouldFailOnError(true)
+                .shouldDoGC(true)
+                .build();
+
+        Collection<RunResult> results = new Runner(opt).run();
+        BenchmarkPerformanceReporter reporter = BenchmarkPerformanceReporter.of(results);
+        reporter.printDetailedPerformanceReport();
+        reporter.printPerformanceRatiosReport();
+    }
+
+    // The JMH samples are the best documentation for how to use it
+    // http://hg.openjdk.java.net/code-tools/jmh/file/tip/jmh-samples/src/main/java/org/openjdk/jmh/samples/
+    @State(Scope.Thread)
+    public static class BenchmarkState {
+        @Param({"10", "1000", "100000"})
+        int numItems;
+
+        Random random = new Random();
```

Comment:
I think this is very important, especially in cases where the resulting data structure will end up looking differently.
This is not a property based test. We **don't** want to have different structures for each run, because then you cannot compare subsequent results.

I would like to stress again: performance tests should really be deterministic!